objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java

```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
        }
    }
    
+   /**
+    * Uploads requested images to VANTIQ. Called when query parameter "operation" is set to "save".
+    * @param request        The parameters sent with the query.
+    * @param replyAddress   The replyAddress used to send a query response.
+    */
+   public void uploadLocalImages(Map<String, ?> request, String replyAddress) {   
+       String imageDir = null;
+       String imageName = null;
+       String imageDate = null;
+       String dateRange = null;
+       List<String> imageDateList = null;
+       ImageUtil imageUtil;
+       
+       // ------------------------ GATHERING PARATMERS ---------------------------
+       
+       // Checking to make sure there is an image directory specified. (Required)
+       if (request.get("imageDir") instanceof String) {
+           imageDir = (String) request.get("imageDir");
+           
+           // Set up ImageUtil properties
+           imageUtil = setupQueryImageUtil(request);
```

Comment:
Personally, I would probably structure this as "detect negatives then deal with it".  So something like
```
if (!imageDir) {
  send error
} else {
   imageDir = ...
}
imageDate = ...
dateRange = ...
imageName = ...

if (imageName) {
  thingsByName = true
} else {
  if (imageDate & dateRange don't work together) {
    sendError
}
// Now we know things are kosher, deal with them....
```

One thing that might be worth considering:  What if we said that imageDate was always a list.  Then, we short circuit a bunch of this.  If it's size is 2, then we know what's up.  Otherwise, we may have to guess.

that said, what if imageDate was always a list of strings.  Usually, these strings are dates.  But, we also allow "none" or "-" or something to indicate that there's no date.  Then, we turn it into a list of Dates (or date strings or whatever) where the "-" means no date (== -infinity or +infinity).  Then we just use that specification to figure out what we're doing.

Something like that may be less complex to explain -- don't know if it would necessarily be easier to implement, but it probably is because we just check for our list & look for special values.  Things other than "-" (which is probably more international than "none") and valid dates would be right out.